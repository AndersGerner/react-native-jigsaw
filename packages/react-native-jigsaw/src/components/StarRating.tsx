import * as React from "react";
import { View, StyleSheet, StyleProp, ViewStyle } from "react-native";
import Icon from "./Icon";
import { withTheme } from "../core/theming";
import theme from "../styles/DefaultTheme";

type Props = {
  starSize?: number;
  maxStars?: number;
  rating?: number;
  theme: typeof theme;
  style?: StyleProp<ViewStyle>;
};

const StarRating: React.FC<Props> = ({
  starSize = 16,
  maxStars = 5,
  rating = 0,
  theme: { colors },
  style,
}) => {
  const ratingRounded = Math.round(rating * 2) / 2;

  return (
    <View style={[styles.containerStyle, style]}>
      {[...Array(maxStars)].map((_, i) => (
        <Icon
          key={i}
          name={
            ratingRounded - i === 0.5
              ? "MaterialIcons/star-half"
              : "MaterialIcons/star"
          }
          size={starSize}
          color={ratingRounded > i ? colors.primary : colors.divider}
        />
      ))}
    </View>
  );
};

const styles = StyleSheet.create({
  containerStyle: {
    flexDirection: "row",
    alignItems: "center",
  },
});

export default withTheme(StarRating);
