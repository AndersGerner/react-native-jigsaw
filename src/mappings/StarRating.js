import { FORM_TYPES, PROP_TYPES, GROUPS } from "../core/component-types";

export const SEED_DATA_PROPS = {
  starSize: {
    label: "Star size",
    description: "Size of each individual star",
    formType: FORM_TYPES.number,
    propType: PROP_TYPES.NUMBER,
    defaultValue: 16,
    editable: true,
    required: false,
    group: GROUPS.basic,
  },
  maxStars: {
    label: "Max Stars",
    description: "The max number of stars",
    formType: FORM_TYPES.number,
    propType: PROP_TYPES.NUMBER,
    defaultValue: 5,
    editable: true,
    required: false,
    group: GROUPS.basic,
  },
  rating: {
    label: "Rating",
    description: "The number of stars that should be populated",
    formType: FORM_TYPES.number,
    propType: PROP_TYPES.NUMBER,
    defaultValue: null,
    editable: true,
    required: false,
    group: GROUPS.data,
  },
};
