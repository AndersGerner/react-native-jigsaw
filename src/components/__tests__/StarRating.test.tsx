import * as React from "react";
import renderer from "react-test-renderer";
import StarRating from "../StarRating";

it("renders star rating", () => {
  const tree = renderer.create(<StarRating />).toJSON();

  expect(tree).toMatchSnapshot();
});
