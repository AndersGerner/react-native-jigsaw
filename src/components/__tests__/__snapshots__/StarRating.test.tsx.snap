// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders star rating 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
      undefined,
    ]
  }
/>
`;
